tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1

```diff
@@ -611,6 +562,128 @@ function Invoke-InstallModule
   }
 }
 
+# Invoke any script block with a retry logic
+function Invoke-CommandWithRetry
+{
+  [CmdletBinding()]
+  [OutputType([PSObject])]
+  Param
+  (
+    [Parameter(Mandatory=$true,
+      ValueFromPipelineByPropertyName=$true,
+      Position=0)]
+      [System.Management.Automation.ScriptBlock]
+      $ScriptBlock,
+
+    [Parameter(Position=1)]
+      [ValidateNotNullOrEmpty()]
+      [int]$MaxRetries=3,
+
+    [Parameter(Position=2)]
+      [ValidateNotNullOrEmpty()]
+      [int]$RetryDelay=3
+  )
+
+  Begin
+  {
+    $currentRetry = 1
+    $Success = $False
+  }
+
+  Process
+  {
+    do {
+      try
+      {
+        $result = . $ScriptBlock
+        $success = $true
+        return $result
+      }
+      catch
+      {
+        $currentRetry = $currentRetry + 1
+        if ($currentRetry -gt $MaxRetries) {
+          $PSCmdlet.ThrowTerminatingError($PSitem)
+        }
+        else {
+          Write-verbose -Message "Waiting $RetryDelay second(s) before attempting again"
+          Start-Sleep -seconds $RetryDelay
+        }
+      }
+    } while(-not $Success)
+  }
+}
+
+# Select profile according to scope & create if it doesn't exist
+function Select-Profile
+{ 
+  param([string]$scope)
+  if($scope -eq "AllUsers" -and (-not $script:IsAdmin))
+  {
+    Write-Error "Administrator rights are required to use AllUsers scope. Log on to the computer with an account that has Administrator rights, and then try again, or retry the operation by adding `"-Scope CurrentUser`" to your command. You can also try running the Windows PowerShell session with elevated rights (Run as Administrator). " -Category InvalidArgument -ErrorAction Stop
+  }
+
+  if($scope -eq "AllUsers")
+  {
+    $profilePath = $profile.AllUsersAllHosts
+  }
+  else {
+    $profilePath = $profile.CurrentUserAllHosts
+  }
+  if (-not (Test-Path $ProfilePath))
+  {
+    new-item -path $ProfilePath -itemtype file -force | Out-Null
+  }
+  return $profilePath
+}
+
+# Get the latest version of a module in a profile
+function Get-LatestModuleVersion
+{
+  param ([array]$versions)
+
+  $versionEnum = $versions.GetEnumerator()
+  $toss = $versionEnum.MoveNext()
+  $version = $versionEnum.Current
+  return $version
+}
+
+# Create a script block with modules and requiredversions for use with default parameters
+function Get-ScriptBlock
+{
+  param($armProfile)
+
+  $ProfileMap = (Get-AzProfile)
+  $Modules = ($ProfileMap.$armProfile | Get-Member -MemberType NoteProperty).Name
+
+  # In scriptblock first cmd should be "if" and others "elseif". So handle first element separately
+  $ScriptBlock = @()
+  $firstModule = $modules[0].ToLower()
+  $versions = $ProfileMap.$armProfile.$firstModule
+  $version = Get-LatestModuleVersion -versions $versions
+  $ScriptBlock += "if (`$MyInvocation.Line.ToLower().Contains(`"$firstModule`")) { `"$version`" } `r`n"
```

Comment:
I think this is the wrong approach, because it means we are writing a lot of stuff to the profile that will be difficult to remove, and it is not robust over modules being deleted.  I think instead the right thing is to:

(1) Write a function that takes the versionProfile string and $MyInvocation.Line as parameters, gets the correct version, checks for the module version on disk and returns the appropriate value.  This could be impemented in a function in the bootstrapper module, and the Profile function would just check for the bootstrapper module and call the bootstrapper function, which would make this a lot easier to delete.

(2) write this function to the profile, and just invoke the fucntion in the default parameter

The diea is that, if the user removed modules or even removes the bootstrapper, the function will actually do nothing.